src/cdomains/arrayDomain.ml

```diff
@@ -195,102 +195,118 @@ sig
   val move_if_affected_with_length: ?replace_with_const:bool -> idx option -> Q.ask -> t -> Cil.varinfo -> (Cil.exp -> int option) -> t
 end
 
-module Partitioned (Val: LatticeWithSmartOps) (Idx:IntDomain.Z):SPartitioned with type value = Val.t and type idx = Idx.t =
+module Partitioned (Val: LatticeWithSmartOps) (Idx:IntDomain.Z): SPartitioned with type value = Val.t and type idx = Idx.t =
 struct
-  (* Contrary to the description in Michael's master thesis, abstract values here always have the form *)
-  (* (Expp, (Val, Val, Val)). Expp is top when the array is not partitioned. In these cases all three  *)
-  (* values from Val are identical *)
-  module Expp = ExpDomain
-  module Base = Lattice.Prod3 (Val) (Val) (Val)
-  include Lattice.ProdSimple(Expp) (Base)
+  type t = Joint of Val.t | Partitioned of (CilType.Exp.t * (Val.t * Val.t * Val.t)) [@@deriving eq, ord, hash]
```

Comment:
Just a thought: doesn't the `Joint` case just duplicate the trivial array domain? Even if so, it might be a stretch to try to reuse code between the two.

Reply:
Yes, in some sense it does, though I am bit hesitant to add reuse here. The Trvivial domain is very easy to understand and I would want to avoid complicating it by sharing code with the partitioned domain.

Reply:
I didn't mean to reduce trivial domain to a special case of partitioned, but rather doing these cases at the level of `AttributeConfiguredArrayDomain` where we already have cases between different array domains. If it's not partitioned, then it could switch to trivial at that level, removing the case splitting from partitioned altogether. But achieving that would be more refactoring than necessary for now.